ui/components/app/snaps/snap-ui-address-input/snap-ui-address-input.tsx

```diff
@@ -0,0 +1,184 @@
+
+  const inputRef = useRef<HTMLDivElement>(null);
+
+  const initialValue = getValue(name, form) as string;
+
+  const [value, setValue] = useState(
+    initialValue
+      ? parseCaipAccountId(initialValue as CaipAccountId).address
+      : '',
+  );
+  const [matchedAddressName, setMatchedAddressName] = useState<string | null>(
+    null,
+  );
+
+  const accountsInfo: Record<string, string> = useSelector((state) =>
+    getAccountInfoByCaipChainId(state, chainId),
+  );
+
+  /*
+   * Focus input if the last focused input was this input
+   * This avoids losing the focus when the UI is re-rendered
+   */
+  useEffect(() => {
+    if (inputRef.current && name === focusedInput) {
+      (inputRef.current.querySelector('input') as HTMLInputElement).focus();
+    }
+  }, [inputRef]);
+
+  const getMatchedAddressName = (address: string) => {
+    const normalizedAddress = address.toLowerCase();
+    return accountsInfo[normalizedAddress];
+  };
+
+  const debouncedHandleInputChange = debounce(
+    (event: ChangeEvent<HTMLInputElement>) => {
+      handleInputChange(name, event.target.value ?? null, form);
+    },
+    80,
+  );
+
+  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
+    setValue(event.target.value);
+    const matchedName = getMatchedAddressName(event.target.value);
+    if (matchedName) {
+      setMatchedAddressName(matchedName);
+    }
+    // Debouncing to allow for rapid keystrokes before allowing the snap to react to the changes
+    debouncedHandleInputChange(event);
+  };
+
+  const handleFocus = () => setCurrentFocusedInput(name);
+  const handleBlur = () => setCurrentFocusedInput(null);
+
+  const handleClear = () => {
+    setValue('');
+    setMatchedAddressName(null);
+  };
+
+  const MatchedAccountInfo = () => {
+    return (
+      <Box display={Display.Flex} flexDirection={FlexDirection.Column}>
+        {label && (
+          <Label className={classnames('mm-form-text-field__label')}>
+            {label}
+          </Label>
+        )}
+        <Box
+          display={Display.InlineFlex}
+          backgroundColor={BackgroundColor.backgroundDefault}
+          alignItems={AlignItems.center}
+          borderWidth={1}
+          borderRadius={BorderRadius.LG}
+          borderColor={BorderColor.borderMuted}
+          paddingLeft={4}
+          paddingRight={4}
+          gap={2}
+          style={{ height: '48px' }}
+        >
+          <Jazzicon address={value} diameter={24} />
+          <Box flexDirection={FlexDirection.Column} gap={2}>
+            <Text fontWeight={FontWeight.Medium}>{matchedAddressName}</Text>
+            <Text variant={TextVariant.bodyXs} ellipsis>
+              {value}
+            </Text>
+          </Box>
+          <Icon
+            onClick={handleClear}
+            name={IconName.Close}
+            color={IconColor.infoDefault}
+            style={{ cursor: 'pointer' }}
+          />
+        </Box>
+      </Box>
+    );
+  };
+
+  if (matchedAddressName) {
```

Comment:
Wait, is this component not supposed to handle arbitrary accounts too?

Reply:
Oh I was under the impression it was for the user's own accounts? If arbitrary accounts as well, then that should be solved with using the hook you suggested above.

Reply:
We should have Erik clarify, maybe I had just misunderstood. But if it is for arbitrary accounts, when would we switch to show the "match" state?